docs/assessment.md

```diff
@@ -224,6 +224,121 @@ When calculating a student's grade for a group assessment, PrairieLearn will alw
+
+- Submitting specific questions
+- Viewing specific questions
+- Assigning group roles for other students
+
+Although in most cases each student is expected to take one role, students are allowed to take on multiple roles in some narrow scenarios, such as when users leave and join groups after assessments have started.
+
+To opt-in to custom group roles, group roles must be defined at the root of the `infoAssessment.json` file. For example:
+
+```json
+{
+  "groupRoles": [
+    {
+      "name": "Manager",
+      "minimum": 1,
+      "maximum": 1,
+      "canAssignRoles": true
+    },
+    {
+      "name": "Recorder",
+      "minimum": 1,
+      "maximum": 1
+    },
+    {
+      "name": "Reflector",
+      "minimum": 1,
+      "maximum": 1
+    },
+    {
+      "name": "Contributor"
+    }
+  ]
+}
+```
+
+| Attribute        | Type    | Default | Description                                                                  |
+| ---------------- | ------- | ------- | ---------------------------------------------------------------------------- |
+| `name`           | string  | -       | The name of the role.                                                        |
+| `minimum`        | integer | 0       | The minimum required number of students holding this role in the assessment. |
+| `maximum`        | integer | -       | The maximum required number of students holding this role in the assessment. |
+| `canAssignRoles` | boolean | false   | Allow students with this role to assign roles to other students.             |
+
+Students typically select their roles before starting an assessment, but they can change their roles mid-assessment if needed. As a safeguard against invalid role configurations, PrairieLearn prevents students from viewing questions if a group's role configuration does not meet the instructor's specification.
+
+#### Adding permissions for an assessment
+
+Permissions can be configured at the _assessment_, _zone_, or _question_ level.
+
+The schema for permissions is defined as follows:
+
+```json
+{
+  "canView": ["Manager", "Reflector", "Recorder", "Contributor"],
+  "canSubmit": ["Recorder"]
+}
+```
+
+| Attribute   | Type            | Default | Description                                                     |
+| ----------- | --------------- | ------- | --------------------------------------------------------------- |
+| `canView`   | Array of string | -       | The names of roles that can view this part of the assessment.   |
+| `canSubmit` | Array of string | -       | The names of roles that can submit this part of the assessment. |
+
+Setting either attribute to `[]` (empty array) means that **no role** can view/submit that part. Setting either attribute to `null` means that **every role** can view/submit that part.
```

Comment:
I think this is true, and it's probably helpful to include so people don't think they need to set both these properties everywhere.

```suggestion
Setting either attribute to `[]` (empty array) means that **no role** can view/submit that part. Setting either attribute to `null` or omitting it entirely means that **every role** can view/submit that part.
```

Reply:
I fear that this may cause confusion if this is considered in the context of a question where a default is set for a zone/assessment. I'll rewrite this to make it clearer.

Reply:
I realized shortly after submitting this review that yeah, this behavior depends on what's configured at parent levels. Given that, I think this is fine as-is, though I'm happy to review a revised version if you think you can make it more clear.

Reply:
Rephrased in 956f8c47549aa9d8b45ff351078fc0f49da3fdb3.